src/tlo/dependencies.py

```diff
@@ -120,29 +126,35 @@ def depth_first_search(module):
             dependencies = get_dependencies(
                 module_instance_map[module], module_instance_map.keys()
             )
+            module_instance = get_module_class_map(set())
             for dependency in sorted(dependencies):
                 if dependency not in module_instance_map:
-                    alternatives_with_instances = [
-                        name for name, instance in module_instance_map.items()
-                        if dependency in instance.ALTERNATIVE_TO
-                    ]
-                    if len(alternatives_with_instances) != 1:
-                        message = (
-                            f'Module {module} depends on {dependency} which is '
-                            'missing from modules to register'
-                        )
-                        if len(alternatives_with_instances) == 0:
-                            message += f' as are any alternatives to {dependency}.'
-                        else:
-                            message += (
-                                ' and there are multiple alternatives '
-                                f'({alternatives_with_instances}) so which '
-                                'to use to resolve dependency is ambiguous.'
+                    if auto_register_dependencies:
```

Comment:
Just realised that this might not quite works as expected when an 'alternative' module dependency (like `SimplifiedBirths`)  is included in the registered modules. If for example `SimplifiedBirths` is registered and the other full pregnancy related modules such as `Labour` and `Contraception` are not, then if `auto_register_dependencies` is set to `True` then the full pregnancy related modules will be automatically registered if they are a dependency of any of the other registered modules even though in this situation it is likely the user is expecting to use `SimplifiedBirths` instead. I think to resolve this we just need to move the logic here to something like the following

```Python
if dependency not in module_instance_map:
    alternatives_with_instances = [
        name for name, instance in module_instance_map.items()
        if dependency in instance.ALTERNATIVE_TO
    ]
    if len(alternatives_with_instances) == 0 and auto_register_dependencies:
        module_instance_map[dependency] = module_class_map[dependency](
            resourcefilepath=resourcefilepath
        )
        yield from depth_first_search(dependency)
    elif len(alternatives_with_instances) != 1:
        message = ... # This block should be same as previously
    else:
        yield from depth_first_search(alternatives_with_instances[0])
else:
    yield from depth_first_search(dependency)
```

This will only automatically register the dependency if it is not already registered _and_ there are no alternatives registered.

Reply:
@matt-graham . Just noting also that get dependencies function [here](https://github.com/UCL/TLOmodel/blob/1bcb1cbebd93cadabbd802b66f4530ef94f7db2d/src/tlo/dependencies.py#L126) only consider dependencies declared in `INIT_DEPENDENCIES` and `OPTIONAL_INIT_DEPENDENCIES`. it doesn't including those declared in `ADDITIONAL_DEPENDENCIES` which makes auto registering of dependencies [fail here ](https://github.com/UCL/TLOmodel/blob/1bcb1cbebd93cadabbd802b66f4530ef94f7db2d/src/tlo/dependencies.py#L292C5-L297C10) when you include modules like `HIV` which has additional dependencies. Any ideas? or should we re-do the registering of dependencies [here instead of throwing an error?](https://github.com/UCL/TLOmodel/blob/1bcb1cbebd93cadabbd802b66f4530ef94f7db2d/src/tlo/dependencies.py#L292C5-L297C10)

Reply:
Ah that's a good point, I'd forgotten about the distinction between `INIT_DEPENDENCIES` and `ADDITIONAL_DEPENDENCIES`.

I think then it might make sense to split out the functionality for auto-registering missing dependencies in to a separate function from `topologically_sort_modules` that is called before the `if sort_modules` and `if check_all_dependencies` blocks, that is something like

```Python
        if auto_register_dependencies:
            modules = [
                *modules, 
                *initialise_missing_dependencies(modules, resourcefilepath=self.resourcefilepath)
            ]
        if sort_modules:
            modules = list(topologically_sort_modules(modules))
        if check_all_dependencies:
            check_dependencies_present(modules)
```

with the logic for getting set of missing dependencies factored out from `check_dependencies_present` into a new function `get_missing_dependencies` which can then also be used to implement the `initialise_missing_dependencies` as something like (haven't tested this works)

```Python
def get_missing_dependencies(
    module_instances: Iterable[Module],
    get_dependencies: DependencyGetter = get_all_dependencies,
) -> Set[str]:
    """Get the set of missing required dependencies if any from an iterable of modules.

    :param module_instances: Iterable of ``Module`` subclass instances to get missing 
        dependencies for.
    :param get_dependencies: Callable which extracts the set of dependencies to check
        for from a module instance. Defaults to extracting all dependencies.
    :return: Set of ``Module`` subclass names corresponding to missing dependencies.
    """
    module_instances = list(module_instances)
    modules_present = {type(module).__name__ for module in module_instances}
    modules_present_are_alternatives_to = set.union(
        # Force conversion to set to avoid errors when using set.union with frozenset
        *(set(module.ALTERNATIVE_TO) for module in module_instances)
    )
    modules_required = set.union(
        *(set(get_dependencies(module, modules_present)) for module in module_instances)
    )
    missing_dependencies = modules_required - modules_present
    return (
        missing_dependencies - modules_present_are_alternatives_to
    )


def initialise_missing_dependencies(modules, **module_kwargs):
    """Get list of initialised instances of any missing dependencies for an iterable of modules.

    :param modules: Iterable of ``Module`` subclass instances to get instances of missing 
        dependencies for.
    :param module_kwargs: Any keyword arguments to use when initialising missing
        module dependencies.
    :return: List of ``Module`` subclass instances corresponding to missing dependencies.
    """
    module_class_map = get_module_class_map(set())
    missing_dependencies = get_missing_dependencies(
        module_instances, get_all_dependencies
    )
    return [
        module_class_map[dependency](**module_kwargs)
        for dependency in missing_dependencies:
     ]

def check_dependencies_present(
    module_instances: Iterable[Module],
    get_dependencies: DependencyGetter = get_all_dependencies,
):
    """Check whether an iterable of modules contains the required dependencies.

    :param module_instances: Iterable of ``Module`` subclass instances to check.
    :param get_dependencies: Callable which extracts the set of dependencies to check
        for from a module instance. Defaults to extracting all dependencies.

    :raises ModuleDependencyError: Raised if any dependencies are missing.
    """
    missing_dependencies = get_missing_dependencies(
        module_instances, get_dependencies
    )
    if len(missing_dependencies) > 0:
        raise ModuleDependencyError(
            'One or more required dependency is missing from the module list and no '
            'alternative to this / these modules are available either: '
            f'{missing_dependencies}'
        )
```

We want to add dependencies before topologically sorting so that these are taking in to account when sorting, and for the topological sorting we need to only iterate over the `INIT_DEPENDENCIES` as by construction these don't allow any directed cycles to avoid a infinite recursion, while if we tried to also account for registering any `ADDITIONAL_DEPENDENCIES` in this recursion I think we might run the risk of get stucking in infinite recursions.

Reply:
This is great @matt-graham . Now that we have `get_missing_dependencies` and `initialise_missing_dependencies` functions, do we still need the` check_dependencies_present` function? I was thinking we can do that when testing for these functions in `test_module_dependencies`. I'm I right?

Reply:
With the above implementation, we will also have to think of a way of adding dependencies of missing dependencies i.e when `Hiv` is registered the implementation will capture `Tb` as a missing dependency of `Hiv` BUT will not capture `Epi` as a missing dependency of `Tb`

Reply:
Hi @mnjowe - sorry for the delay responding here. You're absolutely right in your last comment that the current implementation will only capture direct missing dependencies not the indirect ([transitive](https://en.wikipedia.org/wiki/Transitive_dependency)) missing dependencies. To get these, one approach would be to use some form of recursion, for example

```Python
def get_missing_dependencies(
    module_instances: Iterable[Module],
    get_dependencies: DependencyGetter = get_all_dependencies,
) -> Set[str]:
    """Get the set of missing required dependencies if any from an iterable of modules.

    :param module_instances: Iterable of ``Module`` subclass instances to get missing
        dependencies for.
    :param get_dependencies: Callable which extracts the set of dependencies to check
        for from a module instance. Defaults to extracting all dependencies.
    :return: Set of ``Module`` subclass names corresponding to missing dependencies.
    """
    module_instances = list(module_instances)
    modules_present = {type(module).__name__ for module in module_instances}
    modules_present_are_alternatives_to = set.union(
        # Force conversion to set to avoid errors when using set.union with frozenset
        *(set(module.ALTERNATIVE_TO) for module in module_instances)
    )
    module_class_map = get_module_class_map(set())

    def add_required_dependencies(module_name, modules_required):
        modules_required.add(module_name)
        module_class = module_class_map[module_name]
        dependencies = get_dependencies(module_class, modules_present)
        for dependency in dependencies:
            if dependency not in modules_required:
                add_required_dependencies(dependency, modules_required)

    modules_required = set()
    for module in modules_present:
        add_required_dependencies(module, modules_required)

    missing_dependencies = modules_required - modules_present
    return (
        missing_dependencies - modules_present_are_alternatives_to
    )
```

Reply:
This is great @matt-graham , thanks. Would you also look at what I have done already in my last commit to see if we can go by that?